### PR TITLE
Switch to React

### PR DESCRIPTION
**Pull request by dcposch**
The goal is to make the app faster and smoother. As a bonus, React is more standard and more actively maintained than hyperx/virtual-dom/main-loop.

See #725 for motivation

@feross

### CODE REVIEW

**Comment by dcposch** [comment]
## Work left to do
- [ ] Fix all React warnings
- [ ] Test speed improvement
- [ ] Fix the packaging scripts

**Comment by dcposch** [comment]
We should also use React in a more idiomatic way. I ported it with minimum changes, but later on we should convert function calls like `{Header(..)}` into their own React components like `<Header ... />`

**Comment by dcposch** [comment]
## Speed test
* Load [this torrent](https://archive.org/details/huckleberry_mfs_librivox) and expand it
* Open the dev tools, go to Timeline view, start recording
* Scroll for a few seconds, then stop recording

The timeline view shows FPS and breaks down where time is being spent.

### hyperx
Can't really scroll. See #725 
**Computing the vdom:** ~500ms
**Applying the vdom to the actual dom:** didn't measure


### React
With React in production mode (extra checks and warnings disabled with NODE_ENV=production, which is what our users see):
**Computing the vdom:** ~10ms
**Applying the vdom to the actual dom:** ~60ms

This is better, but unfortunately the scrolling is still choppy. First, 70ms is a lot longer than 16ms, so you are freezing for a few frames (compared to 60fps) every time you call update(). Second, even when the DOM is not updating at all, Chrome can't scroll it smoothly.

Below, the orange parts correspond to time spent in JS (our code and React). The purple parts are time spent in Chrome rendering frames of scrolling animation while the DOM isn't changing.

screen shot 2016-07-20 at 10 59 35 pm

In short it's better now, but still not smooth. IMO this PR is a good first step. Next we'll have to make the DOM more lightweight (eg by capping the number of file rows we show).

**Comment by dcposch** [comment]
The dom is just SLOW

I saved the torrent list view as static HTML:

screen shot 2016-07-22 at 5 21 44 pm

A few stats:
- `torrent-details.html` is only 97KB
- There are only 351 rows in the table
- The table layout is really easy to compute
  - `table-layout: fixed` in the CSS, so no autosizing columns
  - The rows have a fixed pixel height
  - The side columns have fixed pixel width. The middle column takes the remaining width

And this is plain HTML + CSS, no Javascript of any kind.

Anyway despite that, when mousing over the list, the `:hover` state is very choppy.

As you can see below it runs at about 10FPS.

screen shot 2016-07-22 at 5 34 07 pm

**Comment by feross** [comment]
@dcposch Here's an idea: remove the background image. It's hugely stretched. I bet that's causing this.

File: src/renderer/controllers/torrent-controller.js [hunk 1]
```diff
@@ -129,8 +129,6 @@ module.exports = class TorrentController {
     //       files which are completed after a video starts to play to be added
     //       dynamically to the list of subtitles.
     // checkForSubtitles()
-
-    dispatch('update')
```
Comment (feross) [inline src/renderer/controllers/torrent-controller.js]: Why was this removed?

File: src/renderer/main.js [hunk 2]
```diff
@@ -52,7 +52,7 @@ State.load(onState)
 // the dock icon and drag+drop.
 function onState (err, _state) {
   if (err) return onError(err)
-  state = _state
+  state = window.state = _state // Make available for easier debugging
```
Comment (feross) [inline src/renderer/main.js]: Interesting. The `window.state = _state` line wasn't required before.

Is this file no longer executing in the top-level scope because of the react build step?

File: src/renderer/main.js [hunk 3]
```diff
@@ -74,24 +74,15 @@ function onState (err, _state) {
...
-
   // Listen for messages from the main process
   setupIpc()
 
   // Calling update() updates the UI given the current state
   // Do this at least once a second to give every file in every torrentSummary
   // a progress bar and to keep the cursor in sync when playing a video
   setInterval(update, 1000)
+  window.requestAnimationFrame(renderIfNecessary)
+  app = ReactDOM.render(<App state={state} />, document.querySelector('#body'))
```
Comment (feross) [inline src/renderer/main.js]: `body` shouldn't need an id. There's only ever one body.

File: src/renderer/main.js [hunk 4]
```diff
@@ -132,21 +123,23 @@ function lazyLoadCast () {
...
-  try {
-    return App(state)
-  } catch (e) {
-    console.log('rendering error: %s\n\t%s', e.message, e.stack)
+// Calls render() to go from state -> UI, then applies to vdom to the real DOM.
+// Runs at 60fps, but only executes when necessary
+var needsRender = 0
+
+function renderIfNecessary () {
+  if (needsRender > 1) console.log('combining %d update() calls into one update', needsRender)
```
Comment (feross) [inline src/renderer/main.js]: If this log gets printed out often, it will be a perf hit and spam the console. We might want to axe it once we're sure this code is working correctly.

File: src/renderer/main.js [hunk 5]
```diff
@@ -302,6 +295,7 @@ function backToList () {
     var contentTag = document.querySelector('.content')
     if (contentTag) contentTag.scrollTop = 0
 
+    // TODO dcposch: is this still required with React?
```
Comment (feross) [inline src/renderer/main.js]: It's probably still required because we're still using a `requestAnimationFrame` loop. But maybe there's a cleaner way to tell react to redraw with `app.setState(state)`.

File: src/renderer/views/app.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        <div key='content' className='content'>{getView(state)}</div>
+        {getModal(state)}
+      </div>
+    )
+
+    return vdom
+  }
+}
+
+function getErrorPopover (state) {
```
Comment (feross) [inline src/renderer/views/app.js]: These helper functions should probably be private methods on the `App` class so that `state` doesn't need to be passed in each time.

File: src/renderer/views/create-torrent-error-page.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
+const React = require('react')
+
+const {dispatcher} = require('../lib/dispatcher')
```
Comment (feross) [inline src/renderer/views/create-torrent-error-page.js]: Is this dispatcher library still the best way to accomplish this? Doesn't react offer something cleaner?

File: src/renderer/views/unsupported-media-modal.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      : err
+    var actionButton = state.modal.vlcInstalled
+      ? (<button className='button-raised' onClick={dispatcher('vlcPlay')}>Play in VLC</button>)
+      : (<button className='button-raised' onClick={() => this.onInstall}>Install VLC</button>)
+    var vlcMessage = state.modal.vlcNotFound
+      ? 'Couldn\'t run VLC. Please make sure it\'s installed.'
+      : ''
+    return (
+      <div>
+        <p><strong>Sorry, we can't play that file.</strong></p>
```
Comment (feross) [inline src/renderer/views/unsupported-media-modal.js]: Look at this GitHub rendering fail.

Let's name all the files with JSX in them with a `.jsx` extension so this doesn't happen in the future?

File: static/main.html [hunk 9]
```diff
@@ -3,9 +3,13 @@
   <head>
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width, initial-scale=1">
-    <link rel="stylesheet" href="main.css" charset="utf-8">
+    <title>WebTorrent Desktop</title>
+    <link rel="stylesheet" href="main.css">
+    <script>
+	require('../build/renderer/main.js')
```
Comment (feross) [inline static/main.html]: Why is this no longer a script tag with an `async` property? I did LOTS of perf testing and found that that was fastest. Is this change required to get React to work?

File: static/main.html [hunk 10]
```diff
@@ -3,9 +3,13 @@
...
+    </script>
   </head>
   <body>
-    <script async src="main.js"></script>
+    <div id='body'></div>
```
Comment (feross) [inline static/main.html]: Why not just tell react to use the `<body>` directly?

File: static/main.html [hunk 11]
```diff
@@ -3,9 +3,13 @@
   <head>
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width, initial-scale=1">
-    <link rel="stylesheet" href="main.css" charset="utf-8">
+    <title>WebTorrent Desktop</title>
```
Comment (feross) [inline static/main.html]: The actual browser window gets set with Electron APIs, but I suppose this is fine

File: static/main.html [hunk 12]
```diff
@@ -3,9 +3,13 @@
...
+    <link rel="stylesheet" href="main.css">
```
Comment (feross) [inline static/main.html]: Why is charset not specified anymore? Was this an intentional change?

File: static/webtorrent.html [hunk 13]
```diff
@@ -14,9 +15,11 @@
         height: 140px;
       }
     </style>
+    <script>
+      require('../build/renderer/webtorrent.js')
```
Comment (feross) [inline static/webtorrent.html]: Same question: why not `<script async ...>`

**Comment by feross** [comment]
Hey @dcposch, this is nice work! 👍 👍 👍 

Never used React before, but curious to see how this goes. It looks pretty nice so far. I left lots of comments and questions inline.

Please address them, then we can merge this. I think we shouldn't release for a few days after this is merged since I wasn't able to compare line-by-line for some of the files. Some of the file diffs just show every line removed, then every line added. :( So we should wait a few days for any bugs to get shaken out.

File: static/webtorrent.html (hunk 13 above)
Comment (dcposch) [inline static/webtorrent.html]: Doing that breaks `require()`. If we want `<script ...>` to work, we either have to:
* Write ugly things like `require('../build/renderer/<etc>')` in the actual source files
* Move `main.html` to be next to `main.js`

Doing it this way lets us separate `src/` for JS sources only and `static/` for HTML, CSS, etc

File: static/main.html (hunk 12 above)
Comment (dcposch) [inline static/main.html]: I modified `main.html` to pass validation: https://html5.validator.nu/

File: static/main.html (hunk 11 above)
Comment (dcposch) [inline static/main.html]: Added b/c HTML5 validator requires it: https://html5.validator.nu/

File: static/main.html (hunk 10 above)
Comment (dcposch) [inline static/main.html]: That's what I did before, but React prints a warning when you do that. (For silly reasons that don't apply to us--apparently some ppl have browser extensions that break React when used on `<body>`--but I also didn't see a better way to turn off the warning.)

File: src/renderer/main.js (hunk 3 above)
Comment (dcposch) [inline src/renderer/main.js]: This is for the wrapper `<div>`, needed to avoid React's warning about using `<body>` directly

File: src/renderer/main.js (hunk 2 above)
Comment (dcposch) [inline src/renderer/main.js]: It's no longer in top-level scope because of the `require()` path change, because I moved `main.html` over to `static/`

**Comment by dcposch** [comment]
@feross I tried removing the background image, but that wasn't the cause of the slowness.

File: src/renderer/main.js (hunk 4 above)
Comment (dcposch) [inline src/renderer/main.js]: Sure, removed.

File: src/renderer/main.js (hunk 5 above)
Comment (dcposch) [inline src/renderer/main.js]: I'll try it and see if it's slower. Reason I haven't done that before to is to allow collapsing multiple `update()` calls into one.

**Comment by feross** [comment]
👍

File: src/renderer/controllers/torrent-controller.js (hunk 1 above)
Comment (dcposch) [inline src/renderer/controllers/torrent-controller.js]: It was causing double update()s.

File: src/renderer/main.js (hunk 3 above)
Comment (dcposch) [inline src/renderer/main.js]: Added a comment in `main.html` to explain

File: static/main.html (hunk 9 above)
Comment (dcposch) [inline static/main.html]: I'll test and post the results below

**Comment by dcposch** [comment]
## Startup time
Checking the logs for the `init` timer.

Measured on a 2016 12" Macbook running Mac OS El Capitan.

### Before (18aadf9d236ab19309cbaead44559d5fdae27b01)
main process: 850ms
renderer: 25oms

### After (b51dada8e1c3db8136ca67909d94943e72645cbf)
main process: 900ms
renderer: 350ms

**Comment by dcposch** [comment]
And scrolling thru the 350 files in Adventures of Huckleberry Finn is fast now!

**Comment by feross** [comment]
👍 👍 👍
